dvc/output/local.py

```diff
@@ -61,6 +61,12 @@ def changed(self):
 
         return self.project.cache.local.changed(self.path_info, self.info)
 
+    def changed_cache(self):
+        if not self.use_cache or not self.md5:
+            return True
```

Comment:
:thinking: but if there's no cache, it should always be considered changed, isn't it?

Reply:
In this case, there could be no cache because this output doesn't use it (-O|--outs-no-cache in dvc run). For such outputs that are not cached, we should make sure that they are not changed. So something like:
```
for out in stage.outs:
    if out.use_cache:
        return out.changed_cache()
    else:
        return out.changed()
```

Reply:
Though out.changed() will check if the cache is changed as well 🙁 What we actually need here is changed() method from the Dependency parent class.

Reply:
I think I'm not following you, @efiop , `output.changed_cache()` is only used by `_already_cached` during the `stage.run()` method to check whether we need to run the command or not; There's already a check for that one in the top of the reproduce method:
 
```python
    def reproduce(self, force=False, dry=False, interactive=False):
        if not self.changed() and not force:
            return None
```

```python
    def changed(self):
        ret = any([self._changed_deps(),
                   self._changed_outs(),
                   self._changed_md5()])
```

The `_already_cached` method only covers the case when the output is cached and the dependencies or the stage file has changed.

I hope this explains a little bit my reasoning

Reply:
For the record: We've discussed during the meeting and came up with this:

```for out in self.outs:
    if out.use_cache:
        out.changed_cache()
    else:
        out.changed()
```

^ should replace `all(not out.changed_cache() for out in self.outs)`